### PR TITLE
Live-Migrate With Config-Drive CD-Rom

### PR DESCRIPTION
**Pull request by fwiesel**
These changes implement the following for live-migrating vms with config-drive cdroms:
- First detach the config-drive forcefully (it is hopefully not read during that time)
- Copy the iso over to the destination datastore
- Migrate the VM as before
- Reattach the cdrom on the VM

On failure, the cdrom is re-attached again

This depends on sapcc/oslo.vmware#23

### CODE REVIEW

**Comment by fwiesel** [comment]
I'll split out the change with moving the serializer out to oslo.vmware

**Comment by fwiesel** [comment]
Now based on #340

File: nova/virt/vmwareapi/vmops.py [hunk 1]
```diff
@@ -2772,7 +2913,133 @@ def live_migration(self, instance, migrate_data, volume_mapping):
...
+                                                   "ReconfigVM_Task", vm_ref,
+                                                   spec=config_spec)
+        task_completed = threading.Event()
+
+        def set_task_completed(gt):
+            task_completed.set()
+
+        wait_for_task = utils.spawn(
+                self._session.wait_for_task, reconfig_task)
+        wait_for_task.link(set_task_completed)
```
Comment (joker-at-work) [inline nova/virt/vmwareapi/vmops.py:3017]: There is no check that the task succeded. Does the linked thread get called if the task failed? If no, we run into an endless loop. If yes, we will continue as if successfull.

File: nova/virt/vmwareapi/vmops.py [hunk 2]
```diff
@@ -2772,7 +2913,133 @@ def live_migration(self, instance, migrate_data, volume_mapping):
...
+                self._session._call_method(self._session.vim, "AnswerVM",
+                                           vm_ref,
+                                           questionId=question.id,
+                                           answerChoice=yes.key)
+            task_completed.wait(1)
+
+    def reconfigure_vm_device_change(self, context, instance, devices):
+        if not devices:
+            return
+        serialized = [vutil.serialize_object(spec) for spec in devices]
```
Comment (joker-at-work) [inline nova/virt/vmwareapi/vmops.py:3038]: Why do we have to serialize the specs if we don't push them over the wire but instead call `vm_util` to apply them to the VC?

File: nova/virt/vmwareapi/vmops.py [hunk 3]
```diff
@@ -2772,7 +2913,133 @@ def live_migration(self, instance, migrate_data, volume_mapping):
...
+            # destination compute-host
+            try:
+                target.reconfigure_vm_device_change(context, instance,
+                                                    new_cdroms)
+            except messaging.RemoteError:
+                # Failing now would stop the instance reflecting
+                # the new host
+                LOG.exception("Failed to reconfiguring cdroms. "
+                              "Swallowing error as we can't go back now",
+                               instance=instance)
```
Comment (joker-at-work) [inline nova/virt/vmwareapi/vmops.py:2936]: But also nobody will ever notice we lost the cdrom unless somebody checks sentry by accident and feels responsible to fix this ...

Can we add an action-event for that part? I know there can be multiple events for a server action, but I don't know if we're able to add them here, but that way it would be possible to record such non-fatal errors and the calling operator could just check via API if everything succeeded.

File: nova/virt/vmwareapi/vmops.py [hunk 4]
```diff
@@ -2772,7 +2913,133 @@ def live_migration(self, instance, migrate_data, volume_mapping):
...
+                if not file_name:
+                    continue
+                if not file_name.endswith(CONFIGDRIVE_NAME):
+                    continue
+                dc_info = ds_util.get_dc_info(self._session, datastore)
+                ds_util.file_delete(self._session, file_name, dc_info.ref)
+            except vexc.VimException:
+                # No reason to error out the instance, as it is working
+                # but we need some visiblity here for the operator
+                LOG.exception("Cannot delete file", instance=instance)
```
Comment (joker-at-work) [inline nova/virt/vmwareapi/vmops.py:2954]: Please tell which file, so it's easier to debug.

File: nova/virt/vmwareapi/vmops.py [hunk 5]
```diff
@@ -2772,7 +2913,133 @@ def live_migration(self, instance, migrate_data, volume_mapping):
...
+                               instance=instance)
+
+    def delete_config_drive_files(self, ctxt, instance, cdroms):
+        for cdrom_spec in cdroms:
+            try:
+                datastore = getattr(cdrom_spec.device.backing, 'datastore')
+                if not datastore:
+                    continue
+                file_name = getattr(cdrom_spec.device.backing, 'fileName')
+                if not file_name:
```
Comment (joker-at-work) [inline nova/virt/vmwareapi/vmops.py:2945]: These `getattr()` will raise `AttributeError`. They need an additional `None` like the others.

File: nova/virt/vmwareapi/vmops.py [hunk 6]
```diff
@@ -4271,3 +4538,18 @@ def _get_instance_config_info(self, context, instance, image_meta=None):
...
+            backing = getattr(cdrom, 'backing', None)
+            if not backing:
+                continue
+
+            file_name = getattr(backing, 'fileName', None)
+            if not file_name:
+                continue
+            if not file_name.endswith(CONFIGDRIVE_NAME):
+                raise exception.MigrationPreCheckError(
+                        reason="Can only migrate configdrive")
```
Comment (joker-at-work) [inline nova/virt/vmwareapi/vmops.py:4555]: Maybe better: "Found non-configdrive CD-ROM. Can only migrate configdrive"

**Review by joker-at-work — COMMENTED** [review]
I can't say why, but this code is hard to follow for me and it looks extremely hacky, too :(

File: nova/virt/vmwareapi/vmops.py [hunk 7]
```diff
@@ -518,6 +521,48 @@ def _fetch_vsphere_image(self, context, vi, image_ds_loc):
...
+            LOG.warning("Get esx cookies failed: %s", e,
+                        instance=instance)
+            if not dc_info:
+                dc_info = ds_util.get_dc_info(session, ds_ref)
+            dc_path = vutil.get_inventory_path(session.vim, dc_info.ref)
+
+            host = self._session._host
+            cookies = session.vim.client.options.transport.cookiejar
+
+        if method == 'PUT':
```
[the remark this answers is not in the document]
Comment (fwiesel) [reply nova/virt/vmwareapi/vmops.py:555]: I would rather change it the other way around, and test for "GET", as it is the only read-method.

File: nova/virt/vmwareapi/vmops.py [hunk 8]
```diff
@@ -2460,6 +2505,25 @@ def _revert_migration_update_disks(self, vm_ref, instance, vmdk,
...
 
+    def prepare_ds_transfer(self, ctxt, request_method=None, ds_ref=None,
+                            path=None):
+        schema = 'https'
+        ds = ds_obj.get_datastore_by_ref(self._session, ds_ref)
+        dc_info = ds_util.get_dc_info(self._session, ds_ref)
+        dc_name = 'ha-datacenter'  # When connecting directly to esxi host
+        hosts = ds.get_connected_hosts(self._session)
+        host_ref = ds.choose_host(hosts)
+        server = vim_util.get_entity_name(self._session, host_ref)
```
[the remark this answers is not in the document]
Comment (fwiesel) [reply nova/virt/vmwareapi/vmops.py:2516]: It is the name of the argument `build_url`, but sure, I'll change it to `hostname` (as in urlib.parse)

File: nova/virt/vmwareapi/vmops.py [hunk 9]
```diff
@@ -2460,6 +2505,25 @@ def _revert_migration_update_disks(self, vm_ref, instance, vmdk,
...
+
+        ds_url = ds.build_url(schema, server, path, datacenter_name=dc_name)
+
+        if request_method != "GET":
+            ds_path = ds.build_path(path)
+            ds_util.mkdir(self._session, ds_path.parent, dc_info.ref)
+
+        ticket = ds_url.get_transfer_ticket(self._session, request_method)
+        return {'ds_url': str(ds_url), 'ticket': ticket}
```
[the remark this answers is not in the document]
Comment (fwiesel) [reply nova/virt/vmwareapi/vmops.py:2525]: I'm not sure if it is more useful. I think it is more simple. My problem with tuples is they rely on the order, which seems to me slightly more brittle than names. But if you feel strongly about, I can change it.

File: nova/virt/vmwareapi/vmops.py (hunk 1 above)
Comment (fwiesel) [reply nova/virt/vmwareapi/vmops.py:3017]: Yes, `set_task_completed` gets called on completion of the task, regardless of the status.
We will continue as successful, and the same error as before this patch will be raised.

File: nova/virt/vmwareapi/vmops.py (hunk 2 above)
Comment (fwiesel) [reply nova/virt/vmwareapi/vmops.py:3038]: Just to get a more compact logging. It becomes obsolete with sapcc/oslo.vmware#26.
The monkey-patch calls the serializer with a different argument, which is more extensive

File: nova/virt/vmwareapi/vmops.py (hunk 3 above)
Comment (fwiesel) [reply nova/virt/vmwareapi/vmops.py:2936]: > But also nobody will ever notice we lost the cd-rom unless somebody checks sentry by accident and feels responsible to fix this ...

Well, we do not lose it. It will "just" not be connected. And if no one notices, then we are fine :).

The reason why it is written that way lies in the fact, that if we raise an exception here,
then Nova will record the VM as being still on the old host, despite the actual migration having succeeded.

I am not too much a fan of logging to another additional location.

I see following ways to improve it:
- I can write a hammer/screw script to check and fix that particular error.
- I can search for a way to transport that information to the post-migration step in the destination node and error out there
  - Either via an API change
  - Backchannel: Instead of configuring it right away, I store the value (in memory) on the destination node and then apply the configuration change in the post-migration step. Reminder: The whole migration task is in memory of the conductor node, so it isn't restart safe either (yet)

File: nova/virt/vmwareapi/vmops.py (hunk 4 above)
Comment (fwiesel) [reply nova/virt/vmwareapi/vmops.py:2954]: Done

File: nova/virt/vmwareapi/vmops.py (hunk 5 above)
Comment (fwiesel) [reply nova/virt/vmwareapi/vmops.py:2945]: Fixed

File: nova/virt/vmwareapi/vmops.py (hunk 6 above)
Comment (fwiesel) [reply nova/virt/vmwareapi/vmops.py:4555]: Done

File: nova/virt/vmwareapi/vmops.py [hunk 10]
```diff
@@ -518,6 +521,48 @@ def _fetch_vsphere_image(self, context, vi, image_ds_loc):
...
+                dc_path, ds_path.rel_path)
+        except Exception as e:
+            LOG.warning("Get esx cookies failed: %s", e,
+                        instance=instance)
+            if not dc_info:
+                dc_info = ds_util.get_dc_info(session, ds_ref)
+            dc_path = vutil.get_inventory_path(session.vim, dc_info.ref)
+
+            host = self._session._host
+            cookies = session.vim.client.options.transport.cookiejar
```
[the remark this answers is not in the document]
Comment (fwiesel) [reply nova/virt/vmwareapi/vmops.py:553]: Sure, as soon as it is back-ported.

File: nova/virt/vmwareapi/vmops.py (hunk 2 above)
Comment (joker-at-work) [reply nova/virt/vmwareapi/vmops.py:3041]: Ah, I'm just not reading things right. We just serialize for logging and then don't use the `serialized` anymore. I somehow saw it used in the `vm_util.reconfigure_vm_device_change()` call below O.o

File: nova/virt/vmwareapi/vmops.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (joker-at-work) [reply nova/virt/vmwareapi/vmops.py:2525]: I don't feel strongly, was just wondering why we return that extra structure if we don't need it. Functions returns tend to be tuples, if they return multiple things.

File: nova/virt/vmwareapi/vmops.py (hunk 3 above)
Comment (joker-at-work) [reply nova/virt/vmwareapi/vmops.py:2936]: I just don't see much value in raising an exception-level log if nobody will take care of that and if it's important for the operator doing the migration. instance-action-events would be more visible and could be checked easier.

hammer/screw is fine and all, but someone will have to call it. if we can check on the migration task that some optional operation didn't succeed, the operator could run the screw command right away and have a look.

so I think this is more about visibility for me than about "solving it right", as I would think the visibility could come at much lower costs.

Comment (fwiesel) [reply nova/virt/vmwareapi/vmops.py:2936]: > hammer/screw is fine and all, but someone will have to call it. if we can check on the migration task that some optional operation didn't succeed, the operator could run the screw command right away and have a look.

My idea is, that the operator migrates via `screw os migrate-vms`, which could do that check automatically. I would simply call the vcenter an check if the cdrom is properly attached as part of the whole post-live-migration check already in-place, which reports then the error to the user.

I am reluctant to log things to the database, as I consider it bad style... but it seems that ship has sailed, as it is implemented quite conveniently:
That change should do the trick:
```python
@compute_utils.wrap_instance_event(prefix="vmwareapi")
def reconfigure_vm_device_change(self, context, instance, devices):
    ....
```

Comment (fwiesel) [reply nova/virt/vmwareapi/vmops.py:2936]: The implied question: Would it be sufficient to report the error to the operator via screw, would you still rather the event being logged to the database?

Comment (fwiesel) [reply nova/virt/vmwareapi/vmops.py:2936]: Sure, I can do it. I just want to clarify it more, because I still don't seem to get it, why that is preferable, and I may be missing something. So it is more for my own understanding.

I believe the only error with enough visibility and acted upon would be one in the VM state, which I consider a reasonable ask. I am only hesitant about that because the solutions are not "nice".

I agree, that any other error, such as the log message, or sentry report requires a second step, and is therefor limited in their visibility. As I understand though, that also applies to the instance-actions and their events. That is what I am thinking I am missing.

I think, one has to call `GET /servers/{server_id}/os-instance-actions/{request_id}` to get the individual events of an action. 
I am clearly biased against that interface for stylistic reasons, and was wondering, if it is a common thing people do.
But as far as I can see in the logs from the three larges regions, it happened only once in the last two weeks with your credentials.

On a side-note: Apparently a go-httpclient (I assume kubernetes) is polling the `os-instance-actions` with `changes-since`.

Comment (joker-at-work) [reply nova/virt/vmwareapi/vmops.py:2936]: According to Nova's docs, clients are supposed to query the `os-instance-actions` to retrieve the state and result of asynchronous tasks. I know that people don't really do that, but I don't know why. I try to educate them every time this comes up.

Sentry messages go away after a short time and also contain just a sample of the events. Therefore, they seem unfit if there are multiple migrations in a day.

Log messages from screw basically go away immediately and the nova-compute logs after a couple of weeks max.

The instance-actions stay around, though, so we can still debug and verify a problem after some time when the customer tries to access the configdrive again, because they rebooted in some maintenance window long time after the migration. Therefore, I like this place to store errors.

**Review by joker-at-work — APPROVED** [review]
